File: web/src/helpers/storage/drive.ts
/*
 * Copyright (c) [2025] SUSE LLC
 *
 * All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, contact SUSE LLC.
 *
 * To contact SUSE LLC about this file by physical or electronic mail, you may
 * find current contact information at www.suse.com.
 */

import { apiModel } from "~/api/storage";
import { data } from "~/types/storage";
import { switchSearched } from "~/helpers/storage/search";
import { copyApiModel } from "~/helpers/storage/api-model";

function addDrive(apiModel: apiModel.Config, data: data.Drive): apiModel.Config {
  apiModel = copyApiModel(apiModel);
  apiModel.drives ||= [];
  apiModel.drives.push(data);

  return apiModel;
}

function deleteDrive(apiModel: apiModel.Config, name: string): apiModel.Config {
  apiModel = copyApiModel(apiModel);
  apiModel.drives = apiModel.drives.filter((d) => d.name !== name);

  return apiModel;
}

function switchToDrive(
  apiModel: apiModel.Config,
  oldName: string,
  drive: data.Drive,
): apiModel.Config {
  return switchSearched(apiModel, oldName, drive.name, "drives");
}

export { addDrive, deleteDrive, switchToDrive };

File: service/lib/agama/storage/config_solvers/search_matchers/drive.rb
# frozen_string_literal: true

# Copyright (c) [2026] SUSE LLC
#
# All Rights Reserved.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of version 2 of the GNU General Public License as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
# more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, contact SUSE LLC.
#
# To contact SUSE LLC about this file by physical or electronic mail, you may
# find current contact information at www.suse.com.

require "agama/storage/config_solvers/search_matchers/base"
require "agama/storage/config_solvers/search_matchers/with_driver"
require "agama/storage/config_solvers/search_matchers/with_filesystem"
require "agama/storage/config_solvers/search_matchers/with_name"
require "agama/storage/config_solvers/search_matchers/with_partitions"
require "agama/storage/config_solvers/search_matchers/with_size"

module Agama
  module Storage
    module ConfigSolvers
      module SearchMatchers
        # Matcher for the conditions of a drive search.
        class Drive < Base
          include WithName
          include WithSize
          include WithDriver
          include WithFilesystem
          include WithPartitions
        end
      end
    end
  end
end
